### PR TITLE
HMD keyboard

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f8572096b12c8db67918242b5bccdc02d3c58d26 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-f857209.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-f857209.dmg

**Comment by RyanDowne** [comment]
One thing I noticed is that there is no haptic pulse when clicking the keys on the overlays but there is when using the keyboard on the web-entity.

**Comment by RyanDowne** [comment]
When clicking on the source URL field, the field becomes hidden behind the keyboard. This might not be the only field this happens to but it's the only one I've seen. I would expect the text field should always be just above the keyboard.

**Comment by ctrlaltdavid** [comment]
Known issues to address include:
-  Keyboard raises and lowers when click on new entity while entity editor window open.
-  When click in a field near the bottom of entity properties, the properties pane doesn't scroll to show selected field above the keyboard.
-  Keyboard key presses should provide haptic feedback. (Currently only happens with Web entity.)
-  Switch to numeric keys when numeric field gains focus.
-  Keyboard should only display in HMD mode.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2933e37a0a59c370502922b273f028e47c66b065 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-2933e37.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-2933e37.dmg

**Comment by ctrlaltdavid** [comment]
Removal of duplicate keyboard probably fixes the following:
- Keyboard raising and lowering when select entity with editor open.
- Tabbing between fields in entity editor can make keyboard never show again.
- Changing focus from entities editor to Web entity not lowering keyboard in editor.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e9a398472a81e9a7f29c1014948d6b930f345a02 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e9a3984.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e9a3984.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e3be34528fb7af7a809f38dd1b73244bf0080326 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e3be345.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e3be345.dmg

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: e3be34528fb7af7a809f38dd1b73244bf0080326 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e3be345.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e3be345.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5e842843c61c3103264939bfd4a7784ef5a6e9e7 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-5e84284.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-5e84284.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ad5d525340756a624e6f67ffa6f9ccb59fc78a83 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-ad5d525.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-ad5d525.dmg

**Comment by ctrlaltdavid** [comment]
Testing notes:
- Keyboard displayed in both desktop and HMD modes, at least for now for the purposes of QA.
- Keyboard should be displayed in all dialogs etc. that have keyboard, including but not limited to:
	Location bar
	Running Scripts
	Settings dialogs
	Attachments dialogs
	Marketplace
	Entities editor
	Browser
	Web entity
	(Are there any dialogs missing a keyboard?)
- Keyboard typically displays only when cursor is inside editable text field.
- Keyboard automatically switches to numeric for numeric fields.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4818ad2615408cbaf758e65fd03e5c618dc40195 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-4818ad2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-4818ad2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3d75188500d7d91a7800269a05e7a882cc91dda1 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-3d75188.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-3d75188.dmg

**Comment by Nex-Pro** [comment]
Tested on win10 with HMD and desktop.

Observations:

ATP asset migration dialog does not show keyboard when text field is active.

Edit -> running scripts -> Open script -> from disk -> File name field does not show keyboard.

Edit -> Open and run script file... -> File name field does not show keyboard.

Edit -> console Perhaps we could use keyboard here ?

@ctrlaltdavid Thanks for implementing the changes.

It is still a fuss to input a URL as a lot of swapping is needed to get the extra chars.

**Comment by ctrlaltdavid** [comment]
No code changes ...

ATP asset migration dialog does not show keyboard when text field is active.
... This is intentional because the file name field is read-only, so you can't type into it.

Edit -> running scripts -> Open script -> from disk -> File name field does not show keyboard.
... Intentional because file name field is read-only.

Edit -> Open and run script file... -> File name field does not show keyboard.
... Intentional because file name field is read-only.

Edit -> console Perhaps we could use keyboard here ?
... Would be nice but unfortunately this dialog isn't a QML dialog (yet) so will leave keyboard-less for now. Similarly for the Script Editor window.

It is still a fuss to input a URL as a lot of swapping is needed to get the extra chars.
... Agreed. I believe the keyboard layout/contents going to be changed in a subsequent round of work, to make URLs easier and add in missing characters like "{" and"}". I have some ideas on how this can be accomplished.

**Comment by RyanDowne** [comment]
@Nex-Pro is this a PASS?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1a745094e24bd83fc589eb7ed7e7077c77406872 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-1a74509.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-1a74509.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2e762f9e788469a456d28731cde09bef666a4b6 available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e2e762f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-e2e762f.dmg

File: interface/resources/qml/ToolWindow.qml [hunk 1]
```diff
@@ -47,92 +48,101 @@ ScrollingWindow {
...
-                property string originalUrl: "";
-
-                WebView {
-                    id: webView;
-                    anchors.fill: parent
+        anchors.left: parent.left
+        anchors.top: parent.top
+
+        TabView {
+            id: tabView;
```
Comment (sethalves) [inline interface/resources/qml/ToolWindow.qml]: stray trailing ;

File: interface/resources/qml/ToolWindow.qml [hunk 2]
```diff
@@ -47,92 +48,101 @@ ScrollingWindow {
...
-                    property alias eventBridgeWrapper: eventBridgeWrapper 
-                    
-                    QtObject {
-                        id: eventBridgeWrapper
-                        WebChannel.id: "eventBridgeWrapper"
-                        property var eventBridge;
+                    property string originalUrl: "";
+
+                    WebView {
+                        id: webView;
```
Comment (sethalves) [inline interface/resources/qml/ToolWindow.qml]: trailing ;

File: interface/resources/qml/controls-uit/ContentSection.qml [hunk 3]
```diff
@@ -109,8 +109,13 @@ Column {
             }
 
             MouseArea {
+                // Events are propogated so that any active control is defocussed.
```
Comment (sethalves) [inline interface/resources/qml/controls-uit/ContentSection.qml]: defocused

File: libraries/gl/src/gl/OffscreenQmlSurface.cpp [hunk 4]
```diff
@@ -581,6 +581,7 @@ bool OffscreenQmlRenderThread::allowNewFrame(uint8_t fps) {
 }
 
 OffscreenQmlSurface::OffscreenQmlSurface() {
+    // moveToThread(qApp->thread());
```
Comment (sethalves) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: I guess we decided this isn't needed?

Comment (ctrlaltdavid) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: I haven't touched that bit of code. I supposed it's still a to-be-decided. Should we or shouldn't we?

Comment (ctrlaltdavid) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: I haven't touched that bit of code. I supposed it's still a to-be-decided. Should we or shouldn't we? ... A: Will remove it

Comment (ZappoMan) [inline libraries/gl/src/gl/OffscreenQmlSurface.cpp]: The OffscreenQMLSurface code is going to get changed real soon- there's a pending PR.

**Comment by hifi-gustavo** [comment]
Build console output for commit: bdff71151dab4b3c7f74ab13c832f0b4062603ad available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-bdff711.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-bdff711.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a1458a26fda1a61917678013f079d2db0c49acec available at: https://gist.github.com/a5e0a7afebde862eefdbe34bdbdd5ec8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-a1458a2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8674-a1458a2.dmg

**Comment by OmegaHeron** [comment]
This breaks certain aspects of placing a model in scene from ATP Browser.

Reproduction case.

Pick any FBX model from ATP Asset Browser.  Use Add To World button.  **Do not** change default value for physics shape.  Finish dialog to add.  Model will appear in scene instantly.  Now repeat but use pulldown menu to select any of the other shape types.  Model **will not** appear in scene - it is added to models.json.gz but corrupted somehow.

Added ticket https://highfidelity.fogbugz.com/f/cases/1896/Adding-model-via-ATP-Browser-Broken

**Pull request merged**